src/FsAutoComplete/CodeFixes/AddTypeAliasToSignatureFile.fs

```diff
@@ -0,0 +1,95 @@
+module FsAutoComplete.CodeFix.AddTypeAliasToSignatureFile
+
+open FSharp.Compiler.Symbols
+open FSharp.Compiler.Syntax
+open FSharp.Compiler.Text
+open FSharp.Compiler.CodeAnalysis
+open FsToolkit.ErrorHandling
+open Ionide.LanguageServerProtocol.Types
+open FsAutoComplete.CodeFix.Types
+open FsAutoComplete
+open FsAutoComplete.LspHelpers
+
+// TODO: add proper title for code fix
+let title = "AddTypeAliasToSignatureFile Codefix"
+
+let fix (getParseResultsForFile : GetParseResultsForFile) : CodeFix =
+    fun (codeActionParams : CodeActionParams) ->
+        asyncResult {
+            // TODO: check if the file even has a signature file
+
+            let fileName = codeActionParams.TextDocument.GetFilePath () |> Utils.normalizePath
+            // The converted LSP start position to an FCS start position.
+            let fcsPos = protocolPosToPos codeActionParams.Range.Start
+            // The syntax tree and typed tree, current line and sourceText of the current file.
+            let! (parseAndCheckResults : ParseAndCheckResults, _line : string, sourceText : IFSACSourceText) =
+                getParseResultsForFile fileName fcsPos
+
+            let typeDefnInfo =
+                (fcsPos, parseAndCheckResults.GetParseResults.ParseTree)
+                ||> ParsedInput.tryPick (fun path node ->
+                    match node with
+                    | SyntaxNode.SynTypeDefn (SynTypeDefn (
+                        typeInfo = SynComponentInfo (longId = [ typeIdent ])
+                        typeRepr = SynTypeDefnRepr.Simple (simpleRepr = SynTypeDefnSimpleRepr.TypeAbbrev _)
+                        range = m)) when (Range.rangeContainsPos m fcsPos) ->
+                        Some (typeIdent, m, path)
+                    | _ -> None
+                )
+
+            match typeDefnInfo with
+            | None -> return []
+            | Some (typeName, mTypeDefn, implPath) ->
+
+            match parseAndCheckResults.TryGetSymbolUseFromIdent sourceText typeName with
+            | None -> return []
+            | Some typeSymbolUse ->
```

Comment:
This makes me feel like we could use a short-circuit/guard style function to reduce the nesting, maybe something like

```fsharp
let guard input = 
  match input with
  | None -> Ok []
  | Some i -> Ok i

let! (typeName, mTypeDefn, implPath) = guard typeDefnInfo
```

thoughts? It could go in the `CodeFix` module where some of our other filters/checks go.

Reply:
I think I get the concept but I can't get that example to work in the `asyncResult` thing.

Reply:
We can punt on making this nicer - though I could see some kind of `codeFix` CE that includes guards and custom operations for 'summoning' the relevant pieces of data for a codefix easily :D